### PR TITLE
feature(backend): adding no-tracing mode support for the BE

### PR DESCRIPTION
**Pull request by xoscar**
This PR enables the BE to support the no-tracing mode by adding a synchronous job processing by creating the root span directly from Tracetest.

## Changes

- For no tracing mode, the poller executor now adds the root span, finishes the trace fetch process, and runs assertions
- For tracing mode, we are now injecting the root span directly into the trace

## Fixes

- #1628 

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

https://www.loom.com/share/f97f9ce5aedd46d0b0c03e58acbea567

### CODE REVIEW

File: server/config/config.go [hunk 1]
```diff
@@ -118,14 +118,26 @@ func (c Config) MaxWaitTimeForTraceDuration() time.Duration {
...
+	}
+
 	return &dataStoreConfig, nil
 }
 
+func (c Config) IsDataStoreConfigured() bool {
+	selectedStore := c.Server.Telemetry.DataStore
+	_, found := c.Telemetry.DataStores[selectedStore]
+
+	return found
```
Comment (schoren) [inline server/config/config.go:138]: We could avoid logic duplication by reusing the `DataStore` method

```suggestion
	ds, err := c.DataStore()
	return err == nil && ds != nil
```

File: server/executor/trace_poller.go [hunk 2]
```diff
@@ -108,11 +111,38 @@ func (tp tracePoller) Stop() {
...
+
+	if tp.isDataStoreConfigured {
+		tp.enqueueJob(job)
+	} else {
+		tp.syncProcessJob(job)
+	}
+}
+
+func (tp tracePoller) syncProcessJob(job PollingRequest) {
+	run := job.run
```
Comment (schoren) [inline server/executor/trace_poller.go:131]: this is not a great go pattern. I'd remove this and directly access the values with the extra `run.`

File: server/model/spans.go [hunk 3]
```diff
@@ -323,3 +160,46 @@ func decodeChildren(parent *Span, children []encodedSpan) ([]*Span, error) {
...
+	span := Span{
+		ID:         IDGen.SpanID(),
+		Name:       "Tracetest Trigger",
+		StartTime:  run.CreatedAt,
+		EndTime:    run.ServiceTriggerCompletedAt,
+		Attributes: Attributes{},
+		Children:   []*Span{},
+	}
+
+	span.setMetadataAttributes()
```
Comment (schoren) [inline server/model/spans.go:201]: is this working correctly? The code is making a call to a reference method, but the var is not a reference

Comment (schoren) [inline server/model/spans.go:201]: I would replace the pointer receivers with a by value calls:


```
func (span Span) setTriggerResultAttributes(run *Run) Span {
  // do things on span...
  return span //this is now updated
}
// ...
span = span.setMetadataAttributes()
```

File: server/model/traces.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+type Trace struct {
+	ID       trace.TraceID
+	RootSpan Span
+	Flat     map[trace.SpanID]*Span `json:"-"`
+}
+
+func NewTrace(traceID string, spans []Span) Trace {
+	spanMap := make(map[string]*Span, 0)
+	for _, span := range spans {
+		spanCopy := span
```
Comment (schoren) [inline server/model/traces.go:21]: With my suggested changes above, you can do this:
```
spanCopy := span.setMetadataAttributes()
spanID := span.ID.String()
spanMap[spanID] = &spanCopy
```

File: server/tracedb/tracedb.go [hunk 5]
```diff
@@ -21,20 +20,23 @@ const (
...
 	Close() error
 }
 
 var ErrInvalidTraceDBProvider = fmt.Errorf("invalid traceDB provider: available options are (jaeger, tempo)")
 
 func New(c config.Config, repository model.RunRepository) (db TraceDB, err error) {
 	selectedDataStore, err := c.DataStore()
+
+	if selectedDataStore == nil && err == nil {
+		return nil, nil
```
Comment (schoren) [inline server/tracedb/tracedb.go:33]: Returning nil here will force the rest of the code to do nil checks before accessing the datastore. Returning a `noopDataStore` is a safer approach. You can copy one from here: https://github.com/kubeshop/tracetest/blob/c78b2737da070ca529d2bd5f6443bae64f60719a/server/tracedb/tracedb.go#L28-L33

**Review by schoren — CHANGES_REQUESTED** [review]
Most of my comments are more personal opinions than facts, so use your judgement.

The only required change is the one of using a noopDataStore, because that can create nil pointer errors

**Comment by xoscar** [comment]
Hey @schoren thanks for the review man! I just pushed some of the cleanup changes 😄

**Review by schoren — APPROVED** [review]
Looks great!

**Pull request merged**